fdbserver/DiskQueue.actor.cpp

```diff
@@ -109,7 +110,35 @@ struct SyncQueue : ReferenceCounted<SyncQueue> {
 	}
 };
 
-class RawDiskQueue_TwoFiles {
+template <typename T>
+class Tracked {
+protected:
+	struct TrackMe : NonCopyable {
+		T* self;
+		TrackMe( T* self ) : self(self) {
+			self->recoveryActorCount.set( self->recoveryActorCount.get() + 1 );
+		}
+		~TrackMe() {
+			self->recoveryActorCount.set( self->recoveryActorCount.get() - 1 );
+		}
+	};
+
+	Future<Void> onSafeToDestruct() {
+		return onSafeToDestruct(this);
+	}
+
+	ACTOR static Future<Void> onSafeToDestruct(Tracked<T>* self) {
+		while ( self->recoveryActorCount.get() ) {
+			wait( self->recoveryActorCount.onChange() );
+		}
+		return Void();
+	}
+
+private:
+	AsyncVar<int> recoveryActorCount;
```

Comment:
This convention also appear in the data distribution code. I'm wondering why we prefer both AsyncVar and a separate int?

Reply:
Evan had explained in person that it's because .set() involves creating, copying, and destroying a promise.  Which, though it's not super expensive, is more work than what one really needs, and can involve a lot of spurious wake ups of waiting code, when all waiters on care about when the async var is zero.